v0.21.4 maintenance release prep

cpu opened the pull request:
## Description

This PR is targeted at the `rel-0.21` branch, which was cut from the [v/0.21.3](https://github.com/rustls/rustls/releases/tag/v%2F0.21.3) tag to make a maintenance release for the missed error export without the breaking changes/unrelated work in `main`.

- [x] `cargo update`.
- [x] `cargo outdated`.
- [x] `cargo test --all-features`.
- [x] Update `version` in `Cargo.toml`.
- [x] `cargo publish --dry-run -p rustls`
- [x] PR desc updated with release headlines

## Proposed release notes headlines:

- `ClientConfig.with_single_cert` is now deprecated, prefer `with_client_auth_cert` instead.
- `Error::CertRevocationListError` enum is now exported.
- `RootCertStore::add_parsable_certificates` now takes a  `impl IntoIterator<Item = impl AsRef<[u8]>>`.
 
### lib: export CertRevocationListError enum.
The top level `Error::InvalidCertRevocationList` was exported, but not its inner `CertRevocationListError` enum. This should be done to match other enums (e.g. the `Error::InvalidCertificate`'s exported `CertificateError` enum).

### error: use automatic link for RFC 5280 ref.

<details>
<summary>Fixes a `cargo doc` warning:</summary>

```
warning: this URL is not a hyperlink
   --> rustls/src/error.rs:427:15
    |
427 |     /// [^1]: https://www.rfc-editor.org/rfc/rfc5280#section-5.3.1
    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use an automatic link instead: `<[https://www.rfc-editor.org/rfc/rfc5280#section-5.3.1>`](https://www.rfc-editor.org/rfc/rfc5280#section-5.3.1%3E%60)
    |
    = note: bare URLs are not automatically turned into clickable links
    = note: `#[warn(rustdoc::bare_urls)]` on by default
```
</details>

### deps: update webpki-roots 0.23.0 -> 0.24.0

Resolves `cargo outdated` findings.

###  client: with_single_cert -> with_client_auth_cert 

This commit renames the `ClientConfig` builder's `with_single_cert` function to be called `with_client_auth_cert`. The old `with_single_cert` function is left as an alias for `with_client_auth_cert` and marked as deprecated to encourage users to switch to the new name.

### Take IntoIterator in add_parsable_certificates()

`RootCertStore::add_parsable_certificates` now takes a `impl IntoIterator<Item = impl AsRef<[u8]>>`.

### RELEASING.md: restructure, use github releases

Updated release instructions.

### Prepare to move release notes to github releases

Removes release history changelog from README to prefer GitHub releases.

### cargo: prepare 0.21.4 release.

Updates Cargo.toml version 0.21.3 -> 0.21.4

djc commented:
I think we should target a long(er)-lived release branch for all future 0.21.x releases, it shouldn't be specifically for 0.21.4.

cpu commented:
> I think we should target a long(er)-lived release branch for all future 0.21.x releases

SGTM, I deleted `rel-0.21.4` and changed the base of this PR to a new branch `rel-0.21`, still cut from the `v/0.21.3` tag.

djc commented on README.md:
```diff
@@ -20,6 +20,10 @@ If you'd like to help out, please see [CONTRIBUTING.md](CONTRIBUTING.md).
 
 ## Release history
 
+* Release 0.21.4 (2023-07-07)
```
I think we cherry-pick the release history removal so we can rely on new-style release notes for this?

cpu replied:
Done.

djc reviewed (commented):
I think we can also bring in the RootCertStore IntoIterator thing, right?

cpu commented:
> I think we can also bring in the RootCertStore IntoIterator thing, right?

Done.

cpu commented:
Published rustls v0.21.4 at registry `crates-io`
